docs/design-proposals/node/Unified-Vertical-Scaling-for-Container-and-VM-Pods.md

```diff
@@ -130,6 +140,15 @@ RestartContainer policy takes precedence over NoRestart policy.
 Additionally, if RestartPolicy is 'Never', ResizePolicy should be set to
 NoRestart in order to pass validation.
 
+It is proposed to allow users to set the MAX resource to better control the behavior of the resource resizing.
+This could be a value or percentage specified by users. prior to this policy setting, the ceiling of resources, such as
+CPU and memory will be set to the allocatable value at the host and rely the proper CGroup to control the VM resource
+consumptions.
```

Comment:
I don't understand why using allocable as max leads to relying on cgroup. Maybe I miss something. Do we need cGroup control when we use "2*requested" for max vCPU, what's the difference here?

Reply:
in terms of possibility to exceed the "max" resource level for the vm. 2x or host allocatable is the same. 
the case here is how to determine the max so our code logic and libvirt won't set resource to exceed the "max" of it. this is the reason it is proposed an extra setting in the spec.

we need this for cgroup max-xxx properties. if set to max as host allotable, it is essentially disables the CG control.

Reply:
Could you clarify what is the motivation to let the user define the MAX?  Even the user sets the MAX, your code still needs to check "allocatable value at the host" at the time point of resizing in order to make sure there is enough resource for the new size, right? Also in quite some scanarios, it is hard for the user to pre-define the MAX. for example, in serverless usage scenarios, the serverless function controller does not know the future traffic pattern and would like to resize to the "max allocable value at the host" when there is a burst of requests. BTW, do you use CGroup to control the resource boundary of the POD during the resizing?

Reply:
I think there are a few case we need to clarify:
1.  should the user be expose the option to control the max? some app or service might just want to throttle when a certain ceiling is reached instead of keep allocated resources. for example, the app have fixed size thread pool, the app wont need more cpus than the pool size in general. or they want to control the behavior of the app by limiting the  resources. form this case, we can default to host allocatable, but let user to set if they want.

2. from our own control in the runtime to setup the CG for the VM. for example, for memory usage, it is intended to set the memory_limit_limits_inbytes to the MAX. if the max is host allocatable, this setting will be useless.

we might logically still hold up with the limits in VM resource, but the logical concept should be there in the code.

Reply:
I have a feeling we might refer to different MAX setting here. the doc has been changed/updated. i explicitly added a section to explain the few settings in terms of memory control in VM and how they affects the design here.

Reply:
FYI. in the updated doc. to avoid misinterpretation, i suggested to use:
MaxBootTimeMemoryAllocation -- this refers to the "memory" setting in libvirt,
MaxExpandableMemoryAllocation -- this refers to the max hotplugmemory a user can do after the vm is configured and started
CurrentMemoryALloction -- this is memory the current VM configured to use.

Now, with the different size of memory device support, MaxExapandableMemory is not a concern anymore.

I won't recommend to tweek the MaxBootTimeMemoryALlocation for the reasons i described in the doc. please review the updated doc and let me know if you have concerns.

Now in terms of if we want to expose a handle to use to set up the MaxbootTimeMemoryALlocation --- i would defer this to later discussion in future work #1 when we build our own memory management component.